crates/ra_assists/src/handlers/number_representation.rs

```diff
@@ -0,0 +1,597 @@
+use ra_syntax::{ast, ast::LiteralKind, AstNode, SmolStr};
+use std::fmt;
+
+use crate::{Assist, AssistCtx, AssistId};
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+enum NumberLiteralType {
+    /// A literal without prefix, '42'
+    Decimal,
+    /// Hexadecimal literal, '0x2A'
+    Hexadecimal,
+    /// Octal literal, '0o52'
+    Octal,
+    /// Binary literal, '0b00101010'
+    Binary,
+}
+
+#[derive(Clone, Debug)]
+struct NumberLiteral {
```

Comment:
This is a pretty and general-purpose utility, so I think it's should live in `ra_synax`. I think the right API would be as follows:

* add `pub struct IntNumber(SyntaxToken);` to `tokens.rs`
* suffix/preffix methods could return `Option<&str>` then. I think ideally we might want even `fn suffix_range() -> Option<TextRange>`, `fn suffix_text() -> Option<&str>`, but its ok to not add anything beyond this PR. 
* ideally, duplicaiton with `LiteralKind` should be removed. Perhaps it could return an enum of tokens, instead of a lowered enum? But that, again, is probably out of scope of this change.

Reply:
The main theme of the API we want is that, as long as possible, we should just give "views" into existing syntax nodes and tokens, rather than converting them into a POD rust types. Sort of the opposite of what you'd do in a traditional compiler.